### PR TITLE
Gh 332 see collaborators

### PR DESCRIPTION
**Pull request by rocky341**
#331 #332 

Hi @t92549 @GCHQDeveloper314 

This PR implements add collaborator and delete collaborator functionality. It also provides an extra table in the UI which shows the collaborators on a graph and allows deletion of collaborators.

### CODE REVIEW

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/handlers/DeploymentHandler.java [hunk 1]
```diff
@@ -304,6 +341,52 @@ public boolean onAutoGafferDestroy(final KubernetesClient kubernetesClient) thro
...
+        Map<String, String> labels = new HashMap<>();
+        try {
+            List<Deployment> deploymentList = kubernetesClient.apps().deployments().inNamespace(workerNamespace).withLabel("app.kubernetes.io/instance", graphId).list().getItems();
+            if (!deploymentList.isEmpty()) {
+                for (final Deployment deployment : deploymentList) {
+                    labels = deployment.getMetadata().getLabels();
+                }
+            }
+            List<GraphCollaborator> graphCollaborators = listGraphCollaborators(labels, graphId);
+            //graphCollaborators.forEach(graphCollaborator -> System.out.println(graphCollaborator.getUserName().toLowerCase()));
```
[the remark this answers is not in the document]
Comment (rocky341) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/handlers/DeploymentHandler.java]: Removed

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/handlers/DeploymentHandler.java [hunk 2]
```diff
@@ -304,6 +341,52 @@ public boolean onAutoGafferDestroy(final KubernetesClient kubernetesClient) thro
...
+        Map<String, String> labels = new HashMap<>();
+        try {
+            List<Deployment> deploymentList = kubernetesClient.apps().deployments().inNamespace(workerNamespace).withLabel("app.kubernetes.io/instance", graphId).withLabel("creator", username).list().getItems();
+            if (!deploymentList.isEmpty()) {
+                for (final Deployment deployment : deploymentList) {
+                    labels = deployment.getMetadata().getLabels();
+                }
+            }
+            List<GraphCollaborator> graphCollaborators = listGraphCollaborators(labels, graphId);
+            //graphCollaborators.forEach(graphCollaborator -> System.out.println(graphCollaborator.getUserName().toLowerCase()));
```
[the remark this answers is not in the document]
Comment (rocky341) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/handlers/DeploymentHandler.java]: Removed

File: gaffer-as-a-service/gaas-ui/src/components/view-collaborators/view-collaborators-table.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    TableRow,
+    Tooltip,
+    Zoom,
+} from "@material-ui/core";
+import DeleteOutlineOutlinedIcon from "@material-ui/icons/DeleteOutlineOutlined";
+
+interface IProps {
+    graphCollaborators: GraphCollaborator[];
+    deleteCollaborator: (graphId: string, username: string) => void;
+    //  refreshTable: () => void;
```
[the remark this answers is not in the document]
Comment (rocky341) [reply gaffer-as-a-service/gaas-ui/src/components/view-collaborators/view-collaborators-table.tsx]: Removed

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/model/GraphCollaborator.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2021-2022 Crown Copyright
```
Comment (GCHQDeveloper314) [inline gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/model/GraphCollaborator.java:2]: Why 2021?

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/services/GetCollaboratorsService.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Autowired
+    private MeterRegistry meterRegistry;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GetCollaboratorsService.class);
+
+    @Timed(value = "getGraphCollaborators.time", description = "Time taken to get collaborators on a graph", percentiles = 0)
+    public List<GraphCollaborator> getGraphCollaborators(final String graphId) throws GaaSRestApiException {
+        meterRegistry.counter("GetGafferService", "action", "get").increment();
+        LOGGER.info("Get all graphs = ");
```
Comment (GCHQDeveloper314) [inline gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/services/GetCollaboratorsService.java:43]: Is there something missing after the equals?

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/services/GetCollaboratorsService.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        return gafferClient.getGraphCollaborators(graphId);
+    }
+
+    @Timed(value = "getGraphCollaboratorsByUsername.time", description = "Time taken to get collaborators on a graph", percentiles = 0)
+    public List<GraphCollaborator> getGraphCollaboratorsByUsername(final String graphId, final String username) throws GaaSRestApiException {
+        meterRegistry.counter("GetGafferService", "action", "get").increment();
+        LOGGER.info("Get all graphs = ");
```
Comment (GCHQDeveloper314) [inline gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/services/GetCollaboratorsService.java:50]: Is there something missing after the equals?

File: gaffer-as-a-service/gaas-ui/src/components/view-collaborators/gauge.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2021-2022 Crown Copyright
```
Comment (GCHQDeveloper314) [inline gaffer-as-a-service/gaas-ui/src/components/view-collaborators/gauge.tsx:2]: Why 2021?

File: gaffer-as-a-service/gaas-ui/src/components/view-collaborators/view-collaborators-table.tsx [hunk 8]
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2021-2022 Crown Copyright
```
Comment (GCHQDeveloper314) [inline gaffer-as-a-service/gaas-ui/src/components/view-collaborators/view-collaborators-table.tsx:2]: Why 2021?

File: gaffer-as-a-service/gaas-ui/src/components/view-collaborators/view-collaborators.tsx [hunk 9]
```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2021-2022 Crown Copyright
```
Comment (GCHQDeveloper314) [inline gaffer-as-a-service/gaas-ui/src/components/view-collaborators/view-collaborators.tsx:2]: Why 2021?

File: gaffer-as-a-service/gaas-ui/src/domain/graph-collaborator.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2021-2022 Crown Copyright
```
Comment (GCHQDeveloper314) [inline gaffer-as-a-service/gaas-ui/src/domain/graph-collaborator.ts:2]: Why 2021?

File: gaffer-as-a-service/gaas-ui/src/rest/repositories/get-all-graph-collaborators-repo.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2021-2022 Crown Copyright
```
Comment (GCHQDeveloper314) [inline gaffer-as-a-service/gaas-ui/src/rest/repositories/get-all-graph-collaborators-repo.ts:2]: Why 2021?

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/model/GraphCollaborator.java (hunk 4 above)
Comment (rocky341) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/model/GraphCollaborator.java]: Amended

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/services/GetCollaboratorsService.java (hunk 6 above)
Comment (rocky341) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/services/GetCollaboratorsService.java]: Amended

File: gaffer-as-a-service/gaas-ui/src/components/view-collaborators/view-collaborators.tsx (hunk 9 above)
Comment (rocky341) [reply gaffer-as-a-service/gaas-ui/src/components/view-collaborators/view-collaborators.tsx]: Amended

File: gaffer-as-a-service/gaas-ui/src/domain/graph-collaborator.ts (hunk 10 above)
Comment (rocky341) [reply gaffer-as-a-service/gaas-ui/src/domain/graph-collaborator.ts]: Amended

File: gaffer-as-a-service/gaas-ui/src/rest/repositories/get-all-graph-collaborators-repo.ts (hunk 11 above)
Comment (rocky341) [reply gaffer-as-a-service/gaas-ui/src/rest/repositories/get-all-graph-collaborators-repo.ts]: Amended

File: gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/handlers/DeploymentHandlerTest.java [hunk 12]
```diff
@@ -463,6 +466,127 @@ void shouldNotAddCollaboratorLabelWhenAddGraphCollaboratorWithUsernameCalledAndU
 
     }
 
+    @Test
+    void shouldDeleteCollaboratorWhenDeleteCollaboratorCalled() throws ApiException {
+        // Given
```
[the remark this answers is not in the document]
Comment (rocky341) [reply gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/handlers/DeploymentHandlerTest.java:471]: All the tests in `DeploymentHanderTest.java` are written in this format. I think if we are to change things it should be a separate ticket/PR which brings all tests in line with given/when/then.

File: gaffer-as-a-service/gaas-ui/test/rest/repositories/delete-collaborator-repo.test.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2021-2022 Crown Copyright
```
[the remark this answers is not in the document]
Comment (rocky341) [reply gaffer-as-a-service/gaas-ui/test/rest/repositories/delete-collaborator-repo.test.ts]: Amended

**Comment by rocky341** [comment]
> Just small comments. Also a question, I couldn't get the middleware.js example ui to work with listing collaborators, is this meant to work?

Good point I don't think middleware.js has been updated to reflect the new changes.

**Comment by rocky341** [comment]
> Just small comments. Also a question, I couldn't get the middleware.js example ui to work with listing collaborators, is this meant to work?

I have added getCollaborators to the middleware so you should see the view collaborator table working now.

**Review by t92549 — APPROVED** [review]

**Comment by rocky341** [comment]
> Ready to merge if @rocky341 and @GCHQDeveloper314 are?

Happy for it to be merged :D

**Review by GCHQDeveloper314 — APPROVED** [review]

**Pull request merged**